src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentials.java

```diff
@@ -342,123 +364,87 @@ long getTokenStaleEpochSeconds() {
+            this.snapshotToken = baseCredentials.getCachedToken();
         }
 
         private Object readResolve() {
-            JenkinsJVM.checkNotJenkinsJVM();
-            synchronized (this) {
-                ch = Channel.currentOrFail();
+            if (!JenkinsJVM.isJenkinsJVM()) {
+                synchronized (this) {
+                    ch = Channel.currentOrFail();
+                    this.setCachedToken(snapshotToken);
+                }
+                // snapshot is only needed during transport
+                snapshotToken = null;
+                return this;
+            } else {
+                GitHubAppCredentials resolved = new GitHubAppCredentials(this);
+                resolved.setCachedToken(this.snapshotToken);
+                return resolved;
             }
-            return this;
-        }
-
-        @NonNull
-        @Override
-        public String getUsername() {
-            return appID;
         }
 
         @Override
-        public Secret getPassword() {
+        void refreshStaleToken() {
             JenkinsJVM.checkNotJenkinsJVM();
-            try {
-                String appInstallationToken;
-                synchronized (this) {
-                    try {
-                        if (cachedToken == null || cachedToken.isStale()) {
-                            LOGGER.log(Level.FINE, "Generating App Installation Token for app ID {0} on agent", appID);
-                            cachedToken = ch.call(new GetToken(tokenRefreshData));
-                            LOGGER.log(Level.FINER, "Retrieved GitHub App Installation Token for app ID {0} on agent", appID);
-                        }
-                    } catch (Exception e) {
-                        if (cachedToken != null && !cachedToken.isExpired()) {
-                            // Requesting a new token failed. If the cached token is not expired, continue to use it.
-                            // This minimizes failures due to occasional network instability,
-                            // while only slightly increasing the chance that tokens will expire while in use.
-                            LOGGER.log(Level.WARNING,
-                                "Failed to generate new GitHub App Installation Token for app ID " + appID + " on agent: cached token is stale but has not expired");
-                            // Logging the exception here caused a security exeception when trying to read the agent logs during testing
-                            // Added the exception to a secondary log message that can be viewed if it is needed
-                            LOGGER.log(Level.FINER, () -> Functions.printThrowable(e));
-                        } else {
-                            throw e;
-                        }
-                    }
-                    appInstallationToken = cachedToken.getToken();
+            LOGGER.log(Level.FINE, "Generating App Installation Token for app ID {0} on agent", getAppID());
```

Comment:
This degrades security as now the agent has access to the private key.

Reply:
To be clear, I think there _is_ a way to use `CredentialsSnapshotTaker` that retains the security benefits of the current code; would be a minor refinement.

Reply:
@jglick 
The private key is in a `Secret`.  Wouldn't it still be safe?  Or is there some magic when serializing `Secret`s?

Reply:
@jglick 
You mean like this #336 ?

Reply:
> has access to the private key

Also there is risk in the agent being able to select a different `apiUrl`: it could trick the controller into sending a “GitHub API” request to a malicious endpoint. I am not sure if it could actually mount a useful attack this way (since the controller would be sending a JWT not the original secret) but it is certainly unnerving and best avoided.